package com.example.exceptions;






package com.example.exceptions;




package com.example.exceptions;


Java Exception Handling

Throwable (Base Class)


Error  -- (JVM/ System Level issues) errors which we cannot handle  ( OutOfMemoryError ), errors cannot be predicted , Usually we cannot 
catch in code , serious issues that may occur suddenly  for(i=0 ; i<n;i++)
An error is a serious issue that arises in the JVM or system environment , 
ususally outside the application control . It is also a throwable object , but it 
indicates problems that are not meant to be caught or handled in normal code by the programmers or Developers

eg: JVM runs out of memory 
Infinite recursion or a loop which causes a stack overflow

These are not recoverable .

Exception  --- Application Level issues -- which is an event that disrupts the normal flow of the program during execution. It is 
an object that represents an error or unexpected situation that can be handled gracefully using 
try , catch and finally blocks  also includes throw and throws key word.

try:=  business logic  --  you think that may be the code is risky
catch := handle immediately the exception after occured in try block 
finally block will execute in both the cases 


Runtime Exceptions  ( Unchecked Exceptions ) 

ArithmeticExceptions :  DivideByZero

IllegalArgumentException : NumberFormatException / IllegalThreadStateException

ClassCastException : It may throw when an object is cast to a class it's not an instance of

IndexOutOfBoundExcept : for invalid retreiving the values from the array or string for a invalid index range

Null Pointer Exception : which is pointing to null reference;


Compile Time Exceptions (Checked Exception)

classNotFoundexception

filenotfoundexception

InterruptedException
instantiation exception

nosuchmethod exception



package com.example.customexceptions;

public class InsufficientFundsException extends Exception {
	
	public InsufficientFundsException(String message)
	{
		
		super(message);
		
	}
	
	

}


____________________________________

package com.example.customexceptions;

public class BankAccount {
	
	private String accountHolder;
	private double balance;
	
	public BankAccount(String accountHolder, double balance) {
		super();
		this.accountHolder = accountHolder;
		this.balance = balance;
	}
	
	// throws keyword used with this method which might throw the exception 
	public void withdraw(double amount) throws InsufficientFundsException
	{
		
		if(amount<0)
		{
			
			throw new IllegalArgumentException(" Cannot withdraw a negative amount");
		}
		if (amount > balance)
		{
			// throw creates and throws the exception
			throw new InsufficientFundsException(" Have insufficient funds !" + balance);
		}
		balance -= amount;
		System.out.println( amount + "Amount debited :"  + " and now balance is : " +  balance );
	}

	public String getAccountHolder() {
		return accountHolder;
	}

	public double getBalance() {
		return balance;
	}
	
	
	

}

______________________________________________


package com.example.customexceptions;

import java.util.Scanner;

public class BankMain {

	public static void main(String[] args) {
		
		Scanner sc = new Scanner(System.in);
		System.out.println("Enter the account Holder name:");
		String name = sc.next();
		System.out.println("Enter the amount to open your account :");
		double bal = sc.nextDouble();
		BankAccount acc = new BankAccount(name, bal);
		
		try
		{
			System.out.println("Enter the amount to withdraw :");
			double amt = sc.nextDouble();
			acc.withdraw(amt); // It will call withdraw method with the parameter value from BankAccount class 
		}
		
		
		catch(InsufficientFundsException e)
		{
		
		System.out.println(e.getMessage());
	}
		
		catch(IllegalArgumentException ex)
		{
			System.out.println(ex.getMessage());
		}
		
		System.out.println("Final Balance in your account is :" + acc.getBalance());

	}

}
